mm2src/mm2_main/src/lp_ordermatch.rs

```diff
@@ -2049,31 +2075,66 @@ impl MakerOrder {
+
+                if is_legacy {
+                    // Legacy mode: use maker's price to calculate rel amount
                     OrderMatchResult::Matched((taker_base_amount.clone(), taker_base_amount * &self.price))
                 } else {
-                    OrderMatchResult::NotMatched
+                    // taker_rel_amount must cover the premium requested by maker
+                    let required_rel_amount =
+                        taker_base_amount * &self.price + self.premium.clone().unwrap_or_default();
+                    if taker_rel_amount >= &required_rel_amount {
+                        // TPU mode: treat buy as a limit order using taker's base and rel amounts
+                        OrderMatchResult::Matched((taker_base_amount.clone(), taker_rel_amount.clone()))
+                    } else {
+                        OrderMatchResult::NotMatched
+                    }
                 }
             },
             TakerAction::Sell => {
                 let ticker_match = (self.base == taker.rel || self.base_orderbook_ticker.as_ref() == Some(&taker.rel))
                     && (self.rel == taker.base || self.rel_orderbook_ticker.as_ref() == Some(&taker.base));
                 let taker_price = taker_base_amount / taker_rel_amount;
+                let premium = self.premium.clone().unwrap_or_default();
 
-                // Calculate the resulting base amount using the Maker's price instead of the Taker's.
-                let matched_base_amount = taker_base_amount / &self.price;
-                let matched_rel_amount = taker_base_amount.clone();
+                // Determine the matched amounts depending on version
+                let (matched_base_amount, matched_rel_amount) = if is_legacy {
+                    // Legacy: calculate the resulting base amount using the Maker's price instead of the Taker's.
+                    (taker_base_amount / &self.price, taker_base_amount.clone())
+                } else {
+                    // For TPU, if the total rel amount from the taker (rel is coin which should be sent by taker during swap)
+                    // is less than or equal to the maker's premium, the trade is not possible
+                    if taker_base_amount <= &premium {
+                        return OrderMatchResult::NotMatched;
+                    }
+                    // Calculate the resulting base amount using the maker's price instead of the taker's.
+                    // The maker wants to "take" an additional portion of rel as a premium,
+                    // so we reduce the base amount the maker gives by (premium / price).
+                    let matched_base_amount = &(taker_base_amount - &premium) / &self.price;
+                    (matched_base_amount, taker_base_amount.clone())
+                };
 
+                // Match if all common conditions are met
                 if ticker_match
                     && matched_base_amount <= self.available_amount()
```

Comment:
nit: for TakerAction::Buy you created variables for similar conditions (base_amount_exceeds, below_min_volume and price_too_low). Why not create similar vars here too?

Reply:
Originally we were checking for matched conditions, but during feature implementation I needed to reverse the logic and check for NotMatched for early return. So I introduced separate variables to minimize confusion with the original OrderMatchResult::Matched style code.

No need to do the same for the original Sell logic, since its is already straightforward and readable.